compositor_render/src/transformations/builtin/node.rs

```diff
@@ -0,0 +1,183 @@
+                0,
+                &params.shader_buffer_content(),
+            );
+            self.input_resolutions = input_resolutions;
+        }
+    }
+}
+
+enum BuiltinNodeState {
+    Constructed(ConstructedBuiltinNode),
+    Configured(ConfiguredBuiltinNode),
+}
+
+pub struct BuiltinNode(BuiltinNodeState);
+
+impl BuiltinNode {
+    pub fn new(shader: Arc<Shader>, spec: Builtin) -> Self {
+        Self(BuiltinNodeState::Constructed(ConstructedBuiltinNode {
+            shader,
+            builtin: spec,
+        }))
+    }
+
+    // Returns Some(Resolution) if output resolution of node can be determined
+    // from spec (on scene update). If output resolution is depended on input resolutions,
+    // then returns None.
+    pub fn resolution_from_spec(&self) -> Option<Resolution> {
+        self.builtin().resolution_from_spec()
+    }
+
+    fn builtin(&self) -> &Builtin {
+        match &self.0 {
+            BuiltinNodeState::Constructed(ConstructedBuiltinNode { builtin, .. }) => builtin,
+            BuiltinNodeState::Configured(ConfiguredBuiltinNode { builtin, .. }) => builtin,
+        }
+    }
+
+    pub fn render(
+        &mut self,
+        sources: &[(&NodeId, &NodeTexture)],
+        target: &mut NodeTexture,
+        pts: Duration,
+    ) {
+        let input_states = Self::input_states(sources);
+        if self.builtin().should_fallback(&input_states) {
+            target.clear();
+            return;
+        }
+
+        let input_resolutions: Vec<Option<Resolution>> = sources
+            .iter()
+            .map(|(_, node_texture)| node_texture.resolution())
+            .collect();
+
+        let configured = self.ensure_configured(input_resolutions);
+
+        let shader = configured.shader.clone();
+
+        let target = target.ensure_size(&shader.wgpu_ctx, configured.output_resolution);
+        shader.render(
+            &configured.params_bind_group,
+            sources,
+            target,
+            pts,
+            configured.clear_color,
+        );
+    }
+
+    fn ensure_configured(
+        &mut self,
+        input_resolutions: Vec<Option<Resolution>>,
+    ) -> &ConfiguredBuiltinNode {
+        match self {
+            BuiltinNode(BuiltinNodeState::Constructed(ref constructed)) => {
+                let configured = ConfiguredBuiltinNode::new(constructed, input_resolutions);
+                *self = BuiltinNode(BuiltinNodeState::Configured(configured));
+            }
+            BuiltinNode(BuiltinNodeState::Configured(configured)) => {
+                configured.ensure_configured(input_resolutions);
+            }
+        };
+
+        match self {
+            BuiltinNode(BuiltinNodeState::Configured(ref configured)) => configured,
+            BuiltinNode(BuiltinNodeState::Constructed(_)) => {
+                unreachable!("Should be configured by previous call")
+            }
+        }
+    }
+
+    fn input_states(inputs: &[(&NodeId, &NodeTexture)]) -> Vec<InputState> {
```

Comment:
If you only need to map state of the texture to some enum returning iterator would be cleaner, but in this case I think it's best to remove this function and accept slice of textures in should_fallback